presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java

```diff
@@ -183,56 +178,30 @@ public Result apply(TableScanNode tableScanNode, Captures captures, Context cont
                 return Result.empty();
             }
 
-            return Result.ofPlanNode(planTableScan(tableScanNode, TRUE_LITERAL, context, metadata, domainTranslator));
+            return Result.ofPlanNode(planTableScan(tableScanNode, TRUE_LITERAL, context.getSession(), context.getSymbolAllocator().getTypes(), context.getIdAllocator(), metadata, parser, domainTranslator));
         }
     }
 
-    private static PlanNode planTableScan(TableScanNode node, Expression predicate, Rule.Context context, Metadata metadata, DomainTranslator domainTranslator)
+    private static PlanNode planTableScan(
+            TableScanNode node,
+            Expression predicate,
+            Session session,
+            TypeProvider types,
+            PlanNodeIdAllocator idAllocator,
+            Metadata metadata,
+            SqlParser parser,
+            DomainTranslator domainTranslator)
     {
-        Expression deterministicPredicate = filterDeterministicConjuncts(predicate);
-        DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.fromPredicate(
+        return AddExchanges.pickTableLayout(
```

Comment:
I'll swap it the other way around for now. I can see your concern, but I don't think a separate class would be much better. After all, it's closely related to these two classes, and not something generic.

I chose to put it in `AddExchange` initially because that was the more up-to-date version. However, `AddExchange` is such a behemoth that having `PickTableLayout` in there can indeed be concerning.